sympy/solvers/solveset.py

```diff
@@ -931,6 +932,9 @@ def solveset(f, symbol=None, domain=S.Complexes):
             raise NotImplementedError(filldedent('''
                 relationship between value and 0 is unknown: %s''' % b))
 
+    if f.has(Max) or f.has(Min):
```

Comment:
Why not just let anything that can be rewritten as Piecewise benefit:

```diff
diff --git a/sympy/solvers/solveset.py b/sympy/solvers/solveset.py
index 1ce7e20..eaeff0c 100644
--- a/sympy/solvers/solveset.py
+++ b/sympy/solvers/solveset.py
@@ -19,6 +19,7 @@
 from sympy.functions import (log, Abs, tan, cot, sin, cos, sec, csc, exp,
                              acos, asin, acsc, asec, arg,
                              piecewise_fold)
+from sympy.functions.elementary.piecewise import Piecewise
 from sympy.functions.elementary.trigonometric import (TrigonometricFunction,
                                                       HyperbolicFunction)
 from sympy.functions.elementary.miscellaneous import real_root
@@ -995,6 +996,16 @@ def solveset(f, symbol=None, domain=S.Complexes):
             raise NotImplementedError(filldedent('''
                 relationship between value and 0 is unknown: %s''' % b))

+    if domain.is_subset(S.Reals):
+        if symbol.is_real is False:
+            return S.EmptySet
+        if symbol.is_real is None:
+            r = Dummy('r', real=True)
+            f = f.xreplace({symbol: r})
+            return solveset(f, r, domain).xreplace({r: symbol})
+    f = f.rewrite(Piecewise)
+
     if isinstance(f, Eq):
         from sympy.core import Add
         f = Add(f.lhs, - f.rhs, evaluate=False)
```
This gives
```
>>> solveset(x*Max(x, 15) - 10, x)
{2/3}
>>> solveset(abs(x)-1,x,S.Reals)
{-1, 1}
>>> solveset(sign(x)-1,x,S.Reals)
Interval.open(0, oo)
```

I'm not sure if the rewrite should occur only if the domain is real, however. If so, it should be part of the `if`-block.

Reply:
@smichr I think the above `if` condition is a good thing to be done.
But `test_issue_9565` and other tests(as I have mentioned above) fails. This is because of rewriting even the `Abs` to `Piecewise`. I think since `Abs` has its own `_solve_abs` equations with `Abs` should not be included in the rewriting part.